packages/ts/react-grid/src/autoform.tsx

```diff
@@ -0,0 +1,80 @@
+import { type AbstractModel, ValidationError, type DetachedModelConstructor } from '@hilla/form';
+import { Button } from '@hilla/react-components/Button.js';
+import { HorizontalLayout } from '@hilla/react-components/HorizontalLayout.js';
+import { VerticalLayout } from '@hilla/react-components/VerticalLayout.js';
+import { useForm } from '@hilla/react-form';
+import { useEffect, useState, type JSX } from 'react';
+import { AutoFormField } from './autoform-field';
+import type { CrudService } from './crud';
+import { getProperties } from './utils';
+
+type SubmitErrorEvent = {
+  error: unknown;
+};
+type SubmitEvent<TItem> = {
+  item: TItem;
+};
+export type AutoFormProps<TItem> = Readonly<{
+  service: CrudService<TItem>;
+  model: DetachedModelConstructor<AbstractModel<TItem>>;
+  item?: TItem;
+  onSubmitError?({ error }: SubmitErrorEvent): void;
+  onSubmit?({ item }: SubmitEvent<TItem>): void;
+}>;
+
+export function AutoForm<TItem>({ service, model, item, onSubmitError, onSubmit }: AutoFormProps<TItem>): JSX.Element {
+  const form = useForm(model, {
+    onSubmit: async (formItem) => service.update(formItem),
+  });
+  const [formError, setFormError] = useState('');
+  useEffect(() => {
+    form.read(item);
+  }, [item]);
+
+  return (
+    <VerticalLayout theme="padding spacing">
+      {getProperties(model).map((propertyInfo) => (
+        <AutoFormField
+          key={`auto-form-field-${propertyInfo.name}`}
+          propertyInfo={propertyInfo}
+          form={form}
+        ></AutoFormField>
+      ))}
+      <div id="formerror" style={{ color: 'red' }}>
+        {formError}
+      </div>
+      <HorizontalLayout theme="spacing">
+        <Button
+          // eslint-disable-next-line @typescript-eslint/no-misused-promises
+          onClick={async () => {
+            try {
+              setFormError('');
+              const newItem = await form.submit();
+              if (!newItem) {
+                // If update returns an empty object, then no update was performed
+                throw new Error('generic error');
+              } else if (onSubmit) {
+                onSubmit({ item: newItem });
+              } else {
+                form.clear();
+              }
```

Comment:
Don't quite understand this part: If there is a callback, then we don't clear the form, otherwise we do.

I think the form should be cleared when:
- There is no `item` and the form has been successfully submitted
- The `item` state changes to null. Might already be covered by `form.read`, not sure though.

Reply:
Yeah the thought was that if you handle submit yourself you might also want to control what happens to the form. But maybe it always should be cleared.

Reply:
Always clearing might be a problem when implementing the CRUD view, I would expect that a selected item stays in the form after saving. But we can also deal with this later.